features/dnsresolver/netty/src/main/java/org/opennms/netmgt/dnsresolver/netty/CaffeineCache.java

```diff
@@ -0,0 +1,138 @@
+ *  1) Limit the cache's size (and intelligently evict entries)
+ *  2) Expose cache stats
+ *
+ * @author jwhite
+ */
+public abstract class CaffeineCache<E> {
+
+    private final Cache<String, Entries<E>> cache;
+
+    /**
+     * Returns {@code true} if this entry should replace all other entries that are already cached for the hostname.
+     */
+    protected abstract boolean shouldReplaceAll(E entry);
+
+    public CaffeineCache(long maxSize) {
+        final Caffeine<String, Entries> caffeine = Caffeine.newBuilder()
+                .expireAfter(new Expiry<String, Entries>() {
+                    @Override
+                    public long expireAfterCreate(String key, Entries value, long currentTime) {
+                        // Use the TTL of the first entry added for the key as the expiry time
+                        return TimeUnit.SECONDS.toNanos(value.ttl());
+                    }
+
+                    @Override
+                    public long expireAfterUpdate(String key, Entries value, long currentTime, long currentDuration) {
+                        // Don't expire after update
+                        return currentDuration;
+                    }
+
+                    @Override
+                    public long expireAfterRead(String key, Entries value, long currentTime, long currentDuration) {
+                        // Don't expire after read
+                        return currentDuration;
+                    }
+                })
+                .recordStats();
+        if (maxSize > 0) {
+            // Apply a size limit, if set
+            caffeine.maximumSize(maxSize);
+        }
+        cache = caffeine.build();
+    }
+
+    public void clear() {
+        cache.invalidateAll();
+    }
+
+    public void clear(String key) {
+        cache.invalidate(key);
+    }
+
+    public Collection<E> get(String key) {
+        final Entries<E> entries = cache.getIfPresent(key);
+        if (entries == null) {
+            return null;
+        }
+        return entries.get();
+    }
+
+    public void cache(String key, E entry, int ttl) {
+        cache.get(key, k -> new Entries<>(ttl)).add(entry, shouldReplaceAll(entry));
+    }
+
+    public long size() {
+        return cache.estimatedSize();
+    }
+
+    public CacheStats stats() {
+        return cache.stats();
+    }
+
+    private static class Entries<E> {
+        private final int ttl;
+        private final CopyOnWriteArraySet<E> container = new CopyOnWriteArraySet<>();
+
+        public Entries(int ttl) {
+            this.ttl = ttl;
+        }
+
+        public int ttl() {
+            return ttl;
+        }
+
+        public void add(E entry, boolean shouldReplaceAll) {
+            if (shouldReplaceAll) {
+                container.clear();
+            }
+            container.add(entry);
+        }
+
+        public Collection<E> get() {
+            return container;
```

Comment:
We should make this immutable to allow manipulation only via the `CaffeineCache` class.

Reply:
I reduced the visibility of this method. We already convert the collection to a new list before returning the results to the caller in the `CaffeineDnsCache` class, so cache users can't directly manipulate the contents of the container.

Reply:
ok.